scripts/apps/ingest/directives/IngestSourcesContent.ts

```diff
@@ -619,15 +620,28 @@ export function IngestSourcesContent(ingestSources, notify, api, $location,
                     return feedingService ? feedingService.templateUrl : '';
                 };
 
+                /**
+                 * Generates ID compatible with MongoDB's ObjectID
+                 */
+                function mongoDBObjectId(): str {
+                    let oid = Math.floor(Date.now()/1000).toString(16);
+                    for (let i = 0; i < 16; i++) {
+                        oid += Math.floor(Math.random()*16).toString(16);
+                    }
+                    return oid;
+                }
+
+
                 /**
                  * Do URL request specified in url_request field
                  * @param provider ingest provider metadata
                  * @param field url_request field metadata
                  */
                 $scope.doUrlRequest = (provider: IProvider, field: IFeedingServiceField): void => {
                     let provider_name = encodeURIComponent(provider.name);
+                    provider.url_id = mongoDBObjectId();
```

Comment:
Why is new `url_id` generated every time and on the client? Couldn't the provider object coming from the server already have that ID present?

Reply:
It's also not good to couple database specific things like mongo ID format to UI code. Client should talk to API and not be coupled to implementation specific things like what kind of database is used.

Reply:
Hi Tomas,

I agree that it's not ideal, but the issue is that at this point we have nothing from backend, the `provider` object is AFAIK created by the client (the `save` button may have not been pressed at all), and we need to identify the HTTP request origin to link it with the provider.

The use of Mongo DB Object ID instead of something like e.g. UUID, is because we can use it directly as an ID in backend, which is indexed. Using something else would involve the creation of a new field, and it was requested in review of the [backend PR review](https://github.com/superdesk/superdesk-core/pull/1941#discussion_r484360976) (cc @petrjasek) to use `_id`.

To make it cleaner, we would need that the client requests a new provider object when `add new` is pressed in ingest settings, and that would involve more work (probably a new endpoint in backend + request by client).

Reply:
`_id` in mongo doesn't have to be ObjectId, as long as it's unique it should be fine

Reply:
@petrjasek really? I was pretty sure it had to be. In this case if I use a uuid here, is it fine?

Reply:
> I agree that it's not ideal, but the issue is that at this point we have nothing from backend, the provider object is AFAIK created by the client (the save button may have not been pressed at all), and we need to identify the HTTP request origin to link it with the provider.

I checked `ingest_providers` collection in the database and all items have `_id`s. Do you want to do authentication when creating a new provider before it has been saved?

Reply:
@tomaskikutis yes they have `_id` once in the DB, but the authentication is done before the provider is saved. When you click on the save button, a test is done to check the settings, which uses the authentication (it basically tries to connect to the mailbox), and if it fails, the save is denied and the provider is not created in backend, thus the URL request is done before anything is created in backend.

Reply:
@tomaskikutis @petrjasek indeed `_id` is working with any string, I've updated the PR to use a UUID so no more implementation specific ID.